C Web Server

VeraSimon opened the pull request:
### Day 1 & 2: server.c
- [x] send_response()
- [x] handle_http_request()
- [x] get_d20()
- [x] get_file()

### Day 3 & 4: cache.c
- [x] cache_put()
- [x] alloc_entry()
- [x] free_entry()
- [x] cache_create()
- [x] cache_get()
- [x] integrate into server.c

### Stretch
- [ ] find_start_of_body()
- [ ] post_save()
- [ ] auto-serve index.html
- [ ] expire cache entries
- [ ] concurrency

VeraSimon committed:
gitignore
Updated .gitignore

VeraSimon committed:
send_response()
Initial send_response implementation

VeraSimon committed:
get_file()
First pass at get_file implementation

VeraSimon committed:
d20
Implemented /d20 endpoint

VeraSimon committed:
get_file()
Fixed issue where app was segfaulting when requested file didnt exist

VeraSimon committed:
get_file()
get_file is now actually returning files as expected, but throws a random segfault after a couple of requests. investigating.

VeraSimon committed:
get_file()
Fixed issue where I was trying to close a NULL file handle when something didnt exist

LuisanSuarez commented on src/server.c:
```diff
@@ -173,17 +173,16 @@ void get_file(int fd, struct cache *cache, char *request_path)
-        printf("Mime type: %s\n", mime_type);
+        printf("Mime type: %s\n\n", mime_type);
         send_response(fd, "HTTP/1.1 200 OK", mime_type, filedata->data, filedata->size);
+        file_free(filedata);
```
nice catch :)

LuisanSuarez commented on src/server.c:
```diff
@@ -67,16 +67,16 @@ int send_response(int fd, char *header, char *content_type, void *body, int cont
+                "Content-Type: %s\n"
+                "\n"
+                "%s",
+                header, asctime(time_info), content_length, content_type, body_content);
```
I mentioned this today but I think having it in writing will help reference later: this will handle text files fine, but it won't work for jpgs or pngs. You'll have to figure out another way to copy the body into the response if you want to do that.

VeraSimon replied:
Oversight in the spec on my part. I thought the image handling was part of the stretch. I’ll get memcpy() added in tomorrow.

LuisanSuarez commented on src/server.c:
```diff
@@ -100,11 +100,26 @@ void get_d20(int fd)
+
+    if (r < 10)
+    {
+        sprintf(rand_num, "0%i", r);
```
jajaja always on!

VeraSimon replied:
Needs a little work so the result isn’t as skewed, but gets the job done for something like this.

LuisanSuarez commented on src/server.c:
```diff
@@ -100,11 +100,26 @@ void get_d20(int fd)
     // IMPLEMENT ME! //
     ///////////////////
 
+    srand(time(NULL));
```
if I may suggest: set this inside of `main(){}`, that way you only seed once per server run instead of once per d20 call. you'll still get different random strings

VeraSimon replied:
I’ve seen some vague allusions to only doing the seed once, but other than possible entropy exhaustion for larger use apps, why do so?

VeraSimon replied:
Updated as suggested, but would still be interested if you know more about this. :)

VeraSimon committed:
send_response()
Updated to allow for sending arbitrary files using memcpy()

VeraSimon committed:
alloc_entry() & free_entry()
Implemented alloc_entry and free_entry

VeraSimon committed:
cache_put()
First attempt at cache_put implementation. No compiler warnings.

VeraSimon committed:
cache_get()
First attempt at cache_get implementation. No compiler warnings or errors.

VeraSimon committed:
cache_create()
First attempt at cache_create implementation. No compiler warnings or errors.

VeraSimon committed:
cache integration
First attempt at cache integration implemented. No related compiler warnings or errors, but cache seems to be returning a some extra unwanted data. Will investigate.

VeraSimon committed:
alloc_entry() & free_entry()
Forgot to initialize prev and next in alloc_entry. Also updated the function to copy the contents of the pointers passed in at initialization. The extra bits being sent went away somewhere in one of those updates.

VeraSimon committed:
bug fixes
Fixed some small bugs in server.c and cache.c.
